### PR TITLE
HADOOP-17287. Support new Instance by non default constructor by ReflectionUtils

### PR DESCRIPTION
**Pull request by maobaolong**
https://issues.apache.org/jira/browse/HADOOP-17287

This PR provide possibility to new instance by non default constructor through ReflectionUtils

### CODE REVIEW

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java [hunk 1]
```diff
@@ -124,15 +124,29 @@ private static void setJobConf(Object theObject, Configuration conf) {
...
+   *
+   * @param theClass class of which an object is created
+   * @param conf Configuration
+   * @param argsType the types of the arguments
+   * @param values the values of the arguments
+   * @return a new object
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T newInstance(Class<T> theClass, Configuration conf,
+      Class<?>[] argsType, Object ... values) {
```
Comment (sunchao) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java:140]: nit: `argsType` -> `argTypes`

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java [hunk 2]
```diff
@@ -124,15 +124,29 @@ private static void setJobConf(Object theObject, Configuration conf) {
...
+  /** Create an object for the given class and initialize it from conf
+   *
+   * @param theClass class of which an object is created
+   * @param conf Configuration
+   * @param argsType the types of the arguments
+   * @param values the values of the arguments
+   * @return a new object
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T newInstance(Class<T> theClass, Configuration conf,
```
Comment (sunchao) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java:139]: Could you add a test in `TestReflectionUtils`?

**Review by sunchao — COMMENTED** [review]
Thanks @maobaolong looks good with two nits.

**Comment by maobaolong** [comment]
@sunchao Thanks for your review, I add a new test and fix the typo issues.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java (hunk 2 above)
Comment (maobaolong) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java:139]: Done

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java (hunk 1 above)
Comment (maobaolong) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java]: Done

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java [hunk 3]
```diff
@@ -124,15 +124,29 @@ private static void setJobConf(Object theObject, Configuration conf) {
...
+   *
+   * @param theClass class of which an object is created
+   * @param conf Configuration
+   * @param argTypes the types of the arguments
+   * @param values the values of the arguments
+   * @return a new object
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T newInstance(Class<T> theClass, Configuration conf,
+      Class<?>[] argTypes, Object ... values) {
```
Comment (sunchao) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java:140]: One more thing we can do here is perhaps checking whether `argTypes.length == values.length`, and throw `IllegalArgumentException` like: "1 parameters are required but 0 arguments are provided".

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestReflectionUtils.java [hunk 4]
```diff
@@ -168,6 +168,13 @@ public void testLogThreadInfo() throws Exception {
         containsString("Process Thread Dump: " + title));
   }
 
+  @Test
+  public void testNewInstanceForNonDefaultConstructor() {
+    Object x = ReflectionUtils.newInstance(
+        NoDefaultCtor.class, null, new Class[] {int.class}, 1);
+    Assert.assertTrue(x instanceof NoDefaultCtor);
```
Comment (sunchao) [inline hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestReflectionUtils.java:175]: I think `Assert` is not imported and compilation will fail. Also can we test the cases when number of parameter and argument doesn't match?

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java (hunk 3 above)
Comment (maobaolong) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/ReflectionUtils.java:140]: Good idea, done.

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestReflectionUtils.java (hunk 4 above)
Comment (maobaolong) [reply hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestReflectionUtils.java]: Thanks for check this for me, done

**Comment by maobaolong** [comment]
@sunchao Thanks for the review and point a lost import for me. Now I add an argument num check and a new test case.

And,  I test the two tests which added by me, both are passed.

Please take another look, thank you.

**Review by sunchao — APPROVED** [review]
+1 - thanks @maobaolong !

**Comment by maobaolong** [comment]
@sunchao Thanks for your review and merge this PR.

**Pull request merged**